src/shell/shell_cmds.cpp

```diff
@@ -2167,3 +2168,46 @@ void DOS_Shell::CMD_VER(char *args)
 		         dos.version.major, dos.version.minor);
 	}
 }
+
+void DOS_Shell::CMD_VOL(char *args)
+{
+	HELP("VOL");
+
+	uint8_t drive_index;
+	size_t len = 0;
+	if (args) {
+		StripSpaces(args);
+		len = strlen(args);
+	}
+	if (len > 0) {
+		if (len < 2 || args[1] != ':') {
+			SyntaxError();
+			return;
+		}
+		drive_index = toupper(*args) - 'A';
+	} else {
+		drive_index = DOS_GetDefaultDrive();
+	}
+	char drive_letter = drive_index + 'A';
+
+	reg_ah = 0x44;            // DOS IOCTL
+	reg_al = 0x0D;            // Generic block device request
+	reg_bl = drive_index + 1; // Drive (0 = default drive, A = 1, B = 2, etc.)  DOSBox drives start A = 0, so add 1.
+	reg_ch = 0x08;            // Device type (0x08 for block device)
+	reg_cl = 0x66;            // Get volume information (returns serial number, volume label, and filesystem type)
+	reg_dx = 0;               // Offset pointer for return information (DS:DX). Set to 0 for beginning of data segment (hopefully this doesn't step on anyone else's memory?)
+	dos.errorcode = 0;
+	CALLBACK_RunRealInt(0x21);
+	if (dos.errorcode) {
+		WriteOut(MSG_Get("SHELL_EXECUTE_DRIVE_NOT_FOUND"), drive_letter);
+		return;
+	}
+	char label[12];
+	uint32_t serial = mem_readd(Segs.phys[ds] + 2);
+	MEM_BlockRead(Segs.phys[ds] + 6, label, 11); // Read in non-null terminated string with spaces at the end.
+	label[11] = 0; // Null terminator
+	unsigned int high = (serial >> 16) & 0xFFFF;
+	unsigned int low = serial & 0xFFFF;
```

Comment:
the right sides become `int`'s, which we know much become 16-bit: suggest using `const auto = check_cast<uint16_t>( bitwise ops )` for both.

Reply:
For that one I specifically wanted `unsigned int` because that's what `printf` expects with the `%X` specifier for hex.  `WriteOut` calls one of the printf variants `std::vsnprintf`.

Reply:
Let's not have `printf` controlling the functional code types; the later should always take precedent.

(There are no issues w/ smaller types of the same sign being passed into larger-type arguments -> `uint8_t` can be passed into `uint64_t` and that's fully compliant; likewise `int8_t` passed into `int` or `int64_t`)

No concerns though if you'd prefer to cast it to `int32_t` to be explicit or for you own knowledge / documentation.

Reply:
I actually didn't know that.  I thought it might invoke undefined behavior like a lot of C functions do if you look at them wrong.  Good to know.  Changed to your suggested auto + check_cast.

Reply:
> it might invoke undefined behavior like a lot of C functions do **_if you look at them wrong._**

:rofl:  so true :-)

---

Oh; on the topic of UBSAN! 

Meson does a great job at giving us these build types without much extra hassle. I've got a couple aliases for:

```shell
meson setup -Dbuildtype=debug -Db_sanitize=address,undefined -Dunit_tests=disabled  build/ubasan

meson compile -C build/ubasan
```

They take quite a while to build and run slowly, but are a great alternative to valgrind (which often feels far slower).